File: spring-boot/src/main/java/springboot/security/CustomAccessDeniedHandler.java
package springboot.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import springboot.common.enums.ErrorCode;
import springboot.domain.dto.AuthDto;

public class CustomAccessDeniedHandler implements AccessDeniedHandler {

    private final ObjectMapper objectMapper = new ObjectMapper();

    @Override
    public void handle(HttpServletRequest request, HttpServletResponse response,
        AccessDeniedException accessDeniedException) throws IOException, ServletException {
        ErrorCode unauthorized = ErrorCode.UNAUTHORIZED;
        response.setContentType("application/json;charset=UTF-8");
        response.setStatus(unauthorized.getHttpCode());

        AuthDto authDto = AuthDto.builder()
                                       .body(new ArrayList<>(List.of("Access Denied.")))
                                       .build();
        String json = this.objectMapper.writeValueAsString(authDto);
        response.getWriter()
                .print(json);
    }
}

File: spring-boot-kt/src/main/kotlin/real/world/springbootkt/global/config/security/CustomAccessDeniedHandler.kt
package real.world.springbootkt.global.config.security

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import real.world.springbootkt.global.common.ErrorCode

class CustomAccessDeniedHandler(private val objectMapper: ObjectMapper) : AccessDeniedHandler {

    override fun handle(
        request: HttpServletRequest, response: HttpServletResponse,
        accessDeniedException: AccessDeniedException
    ) {
        val forbidden: ErrorCode = ErrorCode.FORBIDDEN
        response.apply {
            this.contentType = "application/json;charset=UTF-8"
            this.status = forbidden.httpCode
        }
        val json = objectMapper.writeValueAsString(forbidden)
        response.writer.print(json)
    }
}